src/libraries/Common/src/System/Diagnostics/DiagnosticsHelper.cs

```diff
@@ -35,14 +35,22 @@ internal static bool CompareTags(List<KeyValuePair<string, object?>>? sortedTags
             int size = count / (sizeof(ulong) * 8) + 1;
             BitMapper bitMapper = new BitMapper(size <= 100 ? stackalloc ulong[size] : new ulong[size]);
 
+#if NET9_0_OR_GREATER // ICollection<T> : IReadOnlyCollection<T> on .NET 9+
+            if (tags2 is IReadOnlyCollection<KeyValuePair<string, object?>> tagsCol)
+#else
             if (tags2 is ICollection<KeyValuePair<string, object?>> tagsCol)
+#endif
```

Comment:
Would it make sense to add checks for both IROC and IC types in older TFMs in the interest of consistency?

```suggestion
            if (tags2 is IReadOnlyCollection<KeyValuePair<string, object?>> tagsRoc)
            {
                if (tagsRoc.Count != count)
                {
                    return false;
                }
            }
#if NET9_0_OR_GREATER // ICollection<T> : IReadOnlyCollection<T> on .NET 9+
            if (tags2 is ICollection<KeyValuePair<string, object?>> tagsCol)
            {
                if (tagsCol.Count != count)
                {
                    return false;
                }
            }
#endif
```

Reply:
Hmm. Maybe? It's largely a free gain for .NET 9+, but for older TFMs it has the same tradeoffs that have caused us to not do this previously, namely that most of the relevant collections that implement ICollection also implement IReadOnlyCollection, and most of the enumerables that don't implement ICollection also don't implement IReadOnlyCollection, which means it's likely in the majority case to be an extra interface check for rare gains.

Reply:
There is precedent in us sacrificing performance for older TFMs in order to preserve functional consistency with the bleeding edge. I'm stressing the term functional here since from my perspective this is more about not having divergent behavior when it comes to quirks in user-defined collections (e.g. side-effectful enumeration).

Reply:
In a case like the one highlighted, that leads to non-trivial complexity, as the subsequent code is then doing processing on both of the collections via the selected interface.

In looking through all the changes, this also appears to be the only one where such concerns about enumeration are relevant: the others only use Count as an acceleration rather than using indexing instead of enumeration.

I'd like to skip this for now. If you feel really strongly about it, we can follow-up on this one case.